clients/drcachesim/tools/record_view.cpp

```diff
@@ -0,0 +1,215 @@
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "record_view.h"
+
+#include <cstdio>
+#include <inttypes.h>
+#include <iostream>
+#include <stdint.h>
+#include <string>
+
+#include "memtrace_stream.h"
+#include "trace_entry.h"
+
+namespace dynamorio {
+namespace drmemtrace {
+
+typedef unsigned int uint;
+
+record_analysis_tool_t *
+record_view_tool_create(uint64_t sim_refs)
+{
+    return new dynamorio::drmemtrace::record_view_t(sim_refs);
+}
+
+record_view_t::record_view_t(uint64_t sim_refs)
+    : sim_refs_(sim_refs)
+{
+}
+
+record_view_t::~record_view_t()
+{
+}
+
+bool
+record_view_t::should_skip(void)
+{
+    if (sim_refs_ > 0) {
+        --sim_refs_;
+        return false;
+    }
+    return true;
+}
+
+bool
+record_view_t::parallel_shard_supported()
+{
+    return false;
+}
+
+std::string
+record_view_t::initialize_shard_type(shard_type_t shard_type)
+{
+    return "";
+}
+
+void *
+record_view_t::parallel_shard_init_stream(int shard_index, void *worker_data,
+                                          memtrace_stream_t *shard_stream)
+{
+    return shard_stream;
+}
+
+bool
+record_view_t::parallel_shard_exit(void *shard_data)
+{
+    return true;
+}
+
+std::string
+record_view_t::parallel_shard_error(void *shard_data)
+{
+    return "";
+}
+
+bool
+record_view_t::parallel_shard_memref(void *shard_data, const trace_entry_t &entry)
+{
+    if (should_skip())
+        return true;
+
+    trace_type_t trace_type = (trace_type_t)entry.type;
+    if (trace_type == TRACE_TYPE_INVALID) {
+        std::cerr << "ERROR: trace_entry_t invalid.\n";
+        return false;
+    }
+
+    std::string trace_type_name = std::string(trace_type_names[trace_type]);
+
+    /* Large if-else for all TRACE_TYPE_.  Prints one line per trace_entry_t.
+     * In some cases we use some helper functions (e.g., type_is_instr()), which group
+     * similar TRACE_TYPE_ together, otherwise we compare trace_type against one or two
+     * specific TRACE_TYPE_ directly.
+     */
+    if (trace_type == TRACE_TYPE_HEADER) {
+        trace_version_t trace_version = (trace_version_t)entry.addr;
+        std::string trace_version_name = std::string(trace_version_names[trace_version]);
+        std::cerr << trace_type_name << ", trace_version: " << trace_version_name << "\n";
+    } else if (trace_type == TRACE_TYPE_FOOTER) {
+        std::cerr << trace_type_name << "\n";
+    } else if ((trace_type == TRACE_TYPE_THREAD) ||
+               (trace_type == TRACE_TYPE_THREAD_EXIT)) {
+        uint tid = (uint)entry.addr;
+        std::cerr << trace_type_name << ", tid: " << tid << "\n";
+    } else if (trace_type == TRACE_TYPE_PID) {
+        uint pid = (uint)entry.addr;
+        std::cerr << trace_type_name << ", pid: " << pid << "\n";
+    } else if (trace_type == TRACE_TYPE_MARKER) {
+        /* XXX i#6751: we have a lot of different types of markers; we should use some
```

Comment:
clients/drcachesim/tools/view.cpp already has an exhaustive switch-case. It can be shared with this tool.

Reply:
Ok.
Although not all of it can be shared.
view.cpp does some delayed printing and extra processing, which would defeat the purpose of "record_view": we want one line per trace_entry_t to show the exact sequence of on-disk records, we don't want to do the extra processing of view.cpp, it should be a 1 to 1 "mirror" of the trace file, just human readable.

Reply:
Now sharing (between view and record_view) the printing of all TRACE_MARKER_TYPE_ cases that do not require extra processing.